[NO-JIRA] Add arch doc (#251)
Skyscanner/backpack-android: docs/ARCHITECTURE.md

> ## Backpack-android architecture overview:

Reviewer: ```suggestion
## backpack-android architecture overview:
```

> Backpack-android is a Android a traditional gradle Android lib project, composed of two modules.

Reviewer: ```suggestion
`backpack-android is a traditional Gradle Android lib project, composed of two modules.
```

> Backpack-android is a Android a traditional gradle Android lib project, composed of two modules.

Reviewer 2: ```suggestion
Backpack-android is a traditional gradle Android lib project, composed of two modules.
```

- app: The example app
- Backpack: The Backpack Android components

## App

### Stories

> Stories are defined programmatically in tree like structure that is then used to build the UI.

Reviewer: ```suggestion
Stories are defined programmatically in  tree-like structure that is then used to build the UI.
```

> Stories are defined programmatically in tree like structure that is then used to build the UI.

Reviewer 2: ```suggestion
Stories are defined programmatically in dendroidal structure that is used to build the UI.
```

Author: I have no idea what `dendroidal` means lol. I'll use @shaundon suggestion instead as I think is easier to understand.

Important classes/packages are: 

- **`{javaSrc}/data/ComponentRegistry`**

  Defines a tree of `stories` used to build the UI:

  ```kotlin
  "Chip" story NodeData { ChipStory() },
  "Panel" story NodeData { Story of R.layout.fragment_panel },
  "Spinner" story NodeData({ children -> SubStory of children },
    mapOf(
      "Default" story NodeData { Story of R.layout.fragment_spinner },
      "Small" story NodeData { Story of R.layout.fragment_spinner_small }
    )),
  ```

- **`res/layout/`**

>   Where layouts for each story is defined. They are named `fragment_{story_name}`

Reviewer: ```suggestion
  Where layouts for each story is defined. They are named `fragment_{story_name}`.
```

- **`{javaSrc}/stories`**
  If a story needs dynamic behaviours it will have an accompanying Kotlin class inside the `stories` package.

### Theming

Theming for the example app is set in the `Settings` view and saved in the local shared preferences 
file. 

Important classes/packages are:

- **`{javaSrc}/SettingsActivity`**
  Activity that shows the theming configuration.

- **`{javaSrc}/ThemeApplier`**
  Implementation of a `onActivityCreated` callback that applies the current theme to the activity.

- **`{javaSrc}/BackpackDemoApplication`**
  Where all initialization logic resides, including registering the current theme.

- **`{javaSrc}/data/SharedPreferences`**
  Used to read/write shared preferences for the example app, including the current theme.

## Backpack

> The backpack lib is where all components live. All packages, except `utils` are Backpack components.

Reviewer: ```suggestion
The `backpack` lib is where all components live. All packages, except `utils` are Backpack components.
```

### Component's architecture

There are only a few things that should be observed about the architecture of the components: 

1. Whenever possible the Backpack component should extend from the Android base component and add the additional extra functionality instead of wrapping the Android component inside another `ViewGroup`. This is to keep the view hierarchy as small as possible.
2. All props should be available via XML and programmatically. Whenever a component is available in both XML and code all its properties should also be available in both. 
3. Changing a property programmatically should immediately reflect on the UI.
4. Components should be "extensible", i.e. `open class` in Kotlin. It's also important to think about the component being used that way when writing/reviewing it.

> More can be found here: https://github.com/Skyscanner/backpack-android/tree/master/decisions

Reviewer 2: Maybe this should live in it's own `### decisions` section?

Author: That's a good point. Can we leave it here for now and I can move it do decision files in the future?

### Theming

> - Uses default Android theming mechanism https://developer.android.com/guide/topics/ui/look-and-feel/themes

Reviewer 2: ```suggestion
- Uses the native Android theming mechanism https://developer.android.com/guide/topics/ui/look-and-feel/themes
```

  - E.g. BpkChip can be themed in two ways:
    ```xml
    // app/res/values/themes.xml

    <style name="RedChip">
      <item name="chipSelectedBackgroundColor">@colors/bpkRed500</item>
    </style>
    ```
    1. Individually through the `style` prop:
    ```xml
      <BpkChip style="@style/RedChip" />
    ```
    2. Globally via the `bpkChipStyle` attribute:
    ```xml
    // app/res/values/style.xml
    <style name="RedChip">
      <item name="chipSelectedBackgroundColor">@colors/bpkRed500</item>
    </style>

    <style name="AppTheme" parent="Theme.AppCompat.Light.NoActionBar">
      <item name="bpkChipStyle">@style/RedChip</item>
    </style>
    ```
- Themes can also be set programmatically using a `ContextThemeWrapper`:
  ```Kotlin
  val themedContext = ContextThemeWrapper(context, R.style.RedChip)
  val chip = BpkChip(themedContext)
  ```

### Caveats

The `style` attribute will take precedence over the global `bpkChipStyle`, meaning it is possible to
have a different style for a single component.

It is also possible to set theming properties directly via `xml`:
```xml
  <BpkChip style="@style/RedChip" app:chipSelectedBackgroundColor="@colors/bpkRed500" />
```

> This is not advertised and is not how we intend theming to be used.

Reviewer 2: ```suggestion
This is not advertised and is not a supported way for theming to be used. If you do this, we cannot guarantee that small changes to Backpack will not break your UI.
```

## Testing

There are currently three types of tests in `backpack-android`:

- Connected tests:

>   - Live inside the `Backpack` module and run inside an emulator, hence "connected". This includes mostly unit testing for visual components.

Reviewer: ```suggestion
  - Live inside the `backpack` module and run inside an emulator, hence "connected". This includes mostly unit testing for visual components.
```

Not sure which of capitalised or lower case is better for the module name but it's done in two ways in this file, so my suggestions change it to lower case everywhere. If that isn't right though go for capitalised 😄

Author: well the actual gradle module name is upper case. I'll review it and make them upper case

- Unit tests:

>   - Live inside the `Backpack` module and do not need an emulator to run. This includes all other unit tests that don't require an emulator.

Reviewer: ```suggestion
  - Live inside the `backpack` module and do not need an emulator to run. This includes all other unit tests that don't require an emulator.
```

- Snapshot tests:
  - Live inside the `app` module and run inside an emulator. Some of those tests are very flaky in the 
  CI and will be annotated with `@FlakyTest`. This means they will not run on CI but still run locally before a release.

> 

Reviewer 2: Such line break. Many wow.
